Game Rule documentation (#527)
FabricMC/fabric-docs: develop/game-rules.md

Reviewer: filename should be `game-rules.md`

---
title: Game Rules
description: A guide for adding custom game rules.
authors:
  - Jummit
  - modmuss50
  - Wind292
authors-nogithub:
  - mysterious_dev
  - solacekairos
---

<!---->

::: info PREREQUISITES

You might want to have completed the [translation generation](./data-generation/translations) first, but it is not required.

:::

> Game rules act as world-specific configuration options that the player can change in-game with a command. These variables usually control some function of the world, for example `pvp`, `spawn_monsters`, and `advance_time`.

Reviewer 2: ```suggestion
Game rules act as world-specific configuration options that the player can change in-game with a command. These variables usually control some function of the world, for example `pvp`, `spawn_monsters`, and `advance_time` control whether PvP is enabled, monster spawning, and time passing.
```

## Creating a Game Rule {#creating-a-game-rule}

To create a custom game rule, first create a `GameRules` class; this is where we are going to declare our game rules. Inside this class, declare two constants: a game rule identifier and the rule itself.

> @[code lang=java transcludeWith=:::gameruleClass](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

Reviewer 2: ```suggestion
@[code lang=java transcludeWith=:::gameruleClass](@/reference/latest/src/main/java/com/example/docs/gamerule/ExampleModGamerules.java)
```

The category argument (`.category(GameRuleCategory.MISC)`) determines which category the gamerule falls under in the world creation screen. In this example, we have created a boolean game rule with a default value of `false` and an id of `bad_vision`. The stored values in game rules are not limited to booleans; other valid types include `double`s, `integer`s, and `enum`s.

Example of a game rule storing a double:

> @[code lang=java transcludeWith=:::double](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

Reviewer 2: ```suggestion
@[code lang=java transcludeWith=:::double](@/reference/latest/src/main/java/com/example/docs/gamerule/ExampleModGamerules.java)
```

## Accessing a Game Rule {#accessing-a-game-rule}

Now that we have a game rule and its `Identifier`, you can access it anywhere with the `serverWorld.getGameRules().get(GAMERULE)` method, where the argument to the `.get()` is your game rule constant and not the game rule id.

> @[code lang=java transcludeWith=:::access](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

Reviewer 2: ```suggestion
@[code lang=java transcludeWith=:::access](@/reference/latest/src/main/java/com/example/docs/gamerule/ExampleModGamerules.java)
```

You can also use this to access the values of vanilla game rules

> @[code lang=java transcludeWith=:::vanilla](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

Reviewer 2: ```suggestion
@[code lang=java transcludeWith=:::vanilla](@/reference/latest/src/main/java/com/example/docs/gamerule/ExampleModGamerules.java)
```

For example, for a rule that applies blindness to every player when true, the implementation would be:

> @[code lang=java transcludeWith=:::badvision](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

Reviewer 2: ```suggestion
@[code lang=java transcludeWith=:::badvision](@/reference/latest/src/main/java/com/example/docs/gamerule/ExampleModGamerules.java)
```

## Translations {#translations}

> Now we need to add our gamerule in datagen to give it a display name. So in your English language provider add:

Reviewer 2: ```suggestion
Now, we need to give our game rule a display name so it can be easily understood from the Game Rules screen. To do this via data generation, add the following lines to your language provider:
```

---
title: Gamerules
description: A guide for adding custom gamerules.

> authors:

Reviewer: Does this guide come from the Wiki at all?

Author: I definitely used the wiki as a reference for the implementation, but I wrote the examples and most of the markdown text.

Reviewer: maybe we should credit the wiki authors then

Reviewer: ```yaml
authors:
  - Jummit
  - modmuss50
  - Wind292
authors-nogithub:
  - mysterious_dev
  - solacekairos
```

Author: Okay! I just added them to the credits.

- Booleans
- Doubles
- Integers
- Enums

Example of a gamerule storing a double:

@[code lang=java transcludeWith=:::double](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

## Accessing a Gamerule {#accessing-a-gamerule}

Now that we have a gamerule and its identifier you can access it anywhere with the `serverworld.getGameRules().get(GAMERULE)` method. Where the argument to the `.get()` is your gamerule constant and not the gamerule identifier.

@[code lang=java transcludeWith=:::access](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

You can also use this to access the state of vanilla gamerules

@[code lang=java transcludeWith=:::vanilla](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

For example, for a gamerule that applies blindness to every player when true, the implementation would be:

@[code lang=java transcludeWith=:::badvision](@/reference/latest/src/main/java/com/example/docs/gamerule/ModGamerules.java)

## Translations {#translations}

> Now we need to add our gamerule in datagen to give it a display name. So in your English language provider add:

Reviewer 3: rewrite (grammar)

@[code lang=java transcludeWith=:::gamerule-name](@/reference/latest/src/client/java/com/example/docs/datagen/ExampleModEnglishLangProvider.java)

> Lastly we need to give our gamerule a description. So in your English language provider add:

Reviewer 3: see above

@[code lang=java transcludeWith=:::gamerule-description](@/reference/latest/src/client/java/com/example/docs/datagen/ExampleModEnglishLangProvider.java)

::: info

Both of these mainly just change what is displayed in the world creation menu.

:::

## Changing Gamerules In-Game {#changing-gamerules-in-game}

Now you should be able to change your gamerule in-game with `/gamerule` command, so for our example:

> ``` command

Reviewer: I think this should be ` ```mcfunction `
